third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,885 @@
+        self.constant_folder
+            .rewrite_call(id, oper, args)
+            .map(|exp| self.expand_tuple(exp))
+            .or({
+                // Not completely a constant.
+                // TODO(BUGBUG): consider matching some expressions,
+                // e.g., ((x + c1) + c2) -> (x + (c1 + c2))
+                None
+            })
+    }
+
+    fn rewrite_if_else(&mut self, _id: NodeId, cond: &Exp, then: &Exp, else_: &Exp) -> Option<Exp> {
+        if DISABLE_IF_SIMPLIFICATION {
+            None
+        } else {
+            match cond.as_ref() {
+                ExpData::Value(_, Value::Bool(true)) => Some(then.clone()),
+                ExpData::Value(_, Value::Bool(false)) => Some(else_.clone()),
+                _ => None,
+            }
+        }
+    }
+
+    fn rewrite_sequence(&mut self, id: NodeId, seq: &[Exp]) -> Option<Exp> {
+        if self.eliminate_code {
+            // Check which elements are side-effect-free
+            let mut siter = seq.iter();
+            let _ignore = siter.next_back(); // ignore last element, we have to keep it
+            let mut side_effect_free_elts: Vec<_> = siter
+                .map(|exp| exp.as_ref().is_side_effect_free())
+                .collect();
+            if side_effect_free_elts.iter().all(|elt| *elt) {
+                // All elements other than the last are side-effect free.
+                // (Note that this case includes a singleton sequence.)
+                seq.iter().next_back().cloned()
+            } else if side_effect_free_elts
+                .iter()
+                .any(|elt_is_side_effect_free| *elt_is_side_effect_free)
+            {
+                // At least one element can be removed.
+                side_effect_free_elts.push(false);
+                let new_vec: Vec<_> = zip(seq, side_effect_free_elts)
+                    .filter_map(|(elt, is_side_effect_free)| {
+                        if !is_side_effect_free {
+                            Some(elt.clone())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                Some(ExpData::Sequence(id, new_vec).into_exp())
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    fn rewrite_pattern(&mut self, pat: &Pattern, entering_scope: bool) -> Option<Pattern> {
+        // Note that scope entry patterns are handled in `rewrite_enter_block_scope`, and
+        // `rewrite_lambda`, and `rewrite_block`, if appropriate.
+        if !entering_scope {
+            // This is an Assign, replace vars if needed
+            pat.replace_vars(self.remapped_symbol.borrow_map())
+        } else {
+            None
+        }
+    }
+
+    fn rewrite_lambda(&mut self, id: NodeId, _pat: &Pattern, _body: &Exp) -> Option<Exp> {
+        let loc = self.env.get_node_loc(id);
+        self.env.diag(
```

Comment:
Perhaps we should document this as a prerequisite at the top of the file (i.e., in module documentation).